File: src/execution/runtimeMutationEngine.ts
/**
 * Converted from Python: core/execution/runtime_mutation_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function trackRuntimeMutations(prior: any = null, mutation: any = null): any {
  var mutations: any = [...py.iter(py.or2(prior, () => ([])))];
  if (py.truthy(mutation)) {
    var entry: any = {"kind": py.toStr(py.get(mutation, "kind", "unknown")), "target": py.toStr(py.get(mutation, "target", "")), "tick": py.toInt(py.get(mutation, "tick", 0)), "ordered_index": py.len(mutations)};
    py.listAppend(mutations, entry);
  }
  var sorted_mutations: any = py.sorted(mutations, {key: ((item: any) => [py.toInt(py.get(item, "tick", 0)), py.toInt(py.get(item, "ordered_index", 0)), py.toStr(py.get(item, "kind", ""))]) as (item: any) => any});
  var by_kind: any = {"dom": py.iter(sorted_mutations).filter((m: any) => py.eq(py.get(m, "kind"), "dom")).map((m: any) => m), "native": py.iter(sorted_mutations).filter((m: any) => py.eq(py.get(m, "kind"), "native")).map((m: any) => m), "workflow": py.iter(sorted_mutations).filter((m: any) => py.eq(py.get(m, "kind"), "workflow")).map((m: any) => m), "synchronization": py.iter(sorted_mutations).filter((m: any) => py.eq(py.get(m, "kind"), "sync")).map((m: any) => m), "memory": py.iter(sorted_mutations).filter((m: any) => py.eq(py.get(m, "kind"), "memory")).map((m: any) => m)};
  return {"mutations": sorted_mutations, "by_kind": by_kind, "count": py.len(sorted_mutations), "deterministic_order": true, "bounded": true};
}

File: lib/src/execution/runtime_mutation_engine.dart
/// Port of core/execution/runtime_mutation_engine.track_runtime_mutations.
Map<String, dynamic> trackRuntimeMutations({
  List<Map<String, dynamic>>? prior,
  Map<String, dynamic>? mutation,
}) {
  final List<Map<String, dynamic>> mutations = <Map<String, dynamic>>[
    ...(prior ?? <Map<String, dynamic>>[])
  ];
  if (mutation != null) {
    mutations.add(<String, dynamic>{
      'kind': '${mutation['kind'] ?? 'unknown'}',
      'target': '${mutation['target'] ?? ''}',
      'tick': _asInt(mutation['tick'], 0),
      'ordered_index': mutations.length,
    });
  }

  final List<Map<String, dynamic>> sortedMutations = <Map<String, dynamic>>[
    ...mutations
  ]..sort((Map<String, dynamic> a, Map<String, dynamic> b) {
      final int t = _asInt(a['tick'], 0).compareTo(_asInt(b['tick'], 0));
      if (t != 0) return t;
      final int oi = _asInt(a['ordered_index'], 0)
          .compareTo(_asInt(b['ordered_index'], 0));
      if (oi != 0) return oi;
      return '${a['kind'] ?? ''}'.compareTo('${b['kind'] ?? ''}');
    });

  Map<String, dynamic> byKind = <String, dynamic>{
    'dom': sortedMutations
        .where((Map<String, dynamic> m) => m['kind'] == 'dom')
        .toList(),
    'native': sortedMutations
        .where((Map<String, dynamic> m) => m['kind'] == 'native')
        .toList(),
    'workflow': sortedMutations
        .where((Map<String, dynamic> m) => m['kind'] == 'workflow')
        .toList(),
    'synchronization': sortedMutations
        .where((Map<String, dynamic> m) => m['kind'] == 'sync')
        .toList(),
    'memory': sortedMutations
        .where((Map<String, dynamic> m) => m['kind'] == 'memory')
        .toList(),
  };

  return <String, dynamic>{
    'mutations': sortedMutations,
    'by_kind': byKind,
    'count': sortedMutations.length,
    'deterministic_order': true,
    'bounded': true,
  };
}

int _asInt(dynamic v, int fallback) {
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v) ?? fallback;
  return fallback;
}
